### PR TITLE
SYSENG-1357: generic client default options

### PR DESCRIPTION
**Pull request by anx-mschaefer**
### Description

This PR adds functionality to define default requests options via `WithRequestOptions` when creating a new generic API client.

### Checklist

* [ ] added release notes to `Unreleased` section in [CHANGELOG.md](CHANGELOG.md), if user facing change

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Review by 89Q12 — APPROVED** [review]
LGTM

File: pkg/api/types/options.go [hunk 1]
```diff
@@ -24,33 +24,39 @@ type Options interface {
...
-	ApplyToDestroy(*DestroyOptions)
+	ApplyToDestroy(*DestroyOptions) error
 }
 
+type AnyOption interface {
+	ApplyToAny(Options) error
+}
+
+type Option interface{}
+
```
Comment (LittleFox94) [inline pkg/api/types/options.go:59]: missing doc comments

File: pkg/apis/kubernetes/v1/environments.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package v1
+
+import (
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+)
+
+func EnvironmentProduction(override bool) types.Option {
```
Comment (LittleFox94) [inline pkg/apis/kubernetes/v1/environments.go:8]: missing doc comment

File: pkg/apis/kubernetes/v1/node_pool_test.go [hunk 3]
```diff
@@ -12,7 +12,8 @@ import (
 var _ = Describe("NodePool resource", func() {
 	It("can filter by cluster", func() {
 		np := &NodePool{Cluster: Cluster{Identifier: "foo"}}
-		url, err := np.EndpointURL(types.ContextWithOperation(context.TODO(), types.OperationList))
+		url, err := np.EndpointURL(
+			types.ContextWithOperation(types.ContextWithOptions(context.TODO(), &types.ListOptions{}), types.OperationList))
```
Comment (LittleFox94) [inline pkg/apis/kubernetes/v1/node_pool_test.go:16]: why is this change necessary?

File: pkg/api/options.go [hunk 4]
```diff
@@ -34,3 +38,32 @@ func FullObjects(fullObjects bool) ListOption {
...
+// ApplyToAny applies the EnvironmentOption to any request type
+func (o EnvironmentOption) ApplyToAny(opts types.Options) error {
+	return opts.Set(fmt.Sprintf("environment/%s", o.APIGroup), o.EnvPathSegment, o.Override)
+}
+
+// GetEnvironmentPathSegment retrieves the environment path segment of a given API group
+// or the provided defaultValue if no environment override is set
+func GetEnvironmentPathSegment(ctx context.Context, apiGroup, defaultValue string) (string, error) {
+	if options, err := types.OptionsFromContext(ctx); err != nil {
+		return "", err
```
Comment (LittleFox94) [inline pkg/api/options.go:59]: shouldn't it just return the default value in this case?

File: pkg/api/options.go [hunk 5]
```diff
@@ -34,3 +38,32 @@ func FullObjects(fullObjects bool) ListOption {
...
+	} else if env, err := options.Get(fmt.Sprintf("environment/%s", apiGroup)); errors.Is(err, types.ErrKeyNotSet) {
+		return defaultValue, nil
+	} else if err != nil {
+		return "", err
+	} else if envString, ok := env.(string); !ok {
+		return "", ErrEnvironmentSegmentNotTypeString
```
Comment (LittleFox94) [inline pkg/api/options.go:65]: if the comment above is accepted, this is the last branch that could return an error - could it be sensible to have a `environments` variable on `types.commonOptions` that can only hold strings, removing the need for this function to be able to return an error, which then also helps with the now 5 returns in `apis/kubernetes/v1.endpointURL`?

File: pkg/api/api_implementation.go [hunk 6]
```diff
@@ -79,8 +88,19 @@ func NewAPI(opts ...NewAPIOption) (API, error) {
 // Get the identified object from the engine.
 func (a defaultAPI) Get(ctx context.Context, o types.IdentifiedObject, opts ...types.GetOption) error {
 	options := types.GetOptions{}
+	var err error
+	for _, requestOpt := range a.requestOptions {
+		if getOpt, ok := requestOpt.(types.GetOption); ok {
+			err = errors.Join(err, getOpt.ApplyToGet(&options))
```
Comment (LittleFox94) [inline pkg/api/api_implementation.go:94]: I would've just returned on the first option returning an error, but I guess collecting all of them could be useful in some situations - "ok, would have done differently tho" ^^

**Review by LittleFox94 — CHANGES_REQUESTED** [review]
found a few things that would be worth changing, but please also first note what I wrote you in Mattermost

File: pkg/api/options.go (hunk 4 above)
Comment (anx-mschaefer) [reply pkg/api/options.go:59]: Hmm.. not sure about that. I think Options are always expected to be available on the context when using this helper (`GetEnvironmentPathSegment`) defaulting to prod (in most cases) on error sounds a bit dangerous.

File: pkg/apis/kubernetes/v1/node_pool_test.go (hunk 3 above)
Comment (anx-mschaefer) [reply pkg/apis/kubernetes/v1/node_pool_test.go:16]: See comment below

File: pkg/apis/kubernetes/v1/environments.go (hunk 2 above)
Comment (anx-mschaefer) [reply pkg/apis/kubernetes/v1/environments.go:8]: Added the comment (not pushed yet).

But do we need this file and its content at all in this project? We are already defaulting to the prod environment.

File: pkg/api/api_implementation.go (hunk 6 above)
Comment (anx-mschaefer) [reply pkg/api/api_implementation.go:94]: Did that to make codeclimate less angry (too many returns) :grimacing:

File: pkg/apis/kubernetes/v1/environments.go (hunk 2 above)
Comment (LittleFox94) [reply pkg/apis/kubernetes/v1/environments.go]: hm probably not, good point - up to you

File: pkg/api/options.go (hunk 4 above)
Comment (LittleFox94) [reply pkg/api/options.go:54]: really not sure .. I think the risk to default to prod due to this is pretty low and can be reasonably enough tested with unit tests, but makes the code using this function a lot nicer in exchange

And dealing with the risk is also possible by using tokens that cannot use the wrong service

File: pkg/api/options.go (hunk 5 above)
Comment (LittleFox94) [reply pkg/api/options.go:60]: something like that, yea

File: pkg/api/api_implementation.go [hunk 7]
```diff
@@ -79,8 +88,17 @@ func NewAPI(opts ...NewAPIOption) (API, error) {
...
+		}
+	}
 	for _, opt := range opts {
-		opt.ApplyToGet(&options)
+		err = errors.Join(err, opt.ApplyToGet(&options))
+	}
```
Comment (LittleFox94) [inline pkg/api/api_implementation.go:99]: this can be prettied up a bit for all the request methods with a small generic helper:

```go
func resolveRequestOptions[T any](commonOptions []types.Option, requestOptions []T) []T {
	return append(filterOptions[T](commonOptions), requestOptions...)
}

func filterOptions[T any](opts []types.Option) []T {
	ret := make([]T, 0, len(opts))
	for _, v := range opts {
		if v, ok := v.(T); ok {
			ret = append(ret, v)
		}
	}

	return ret
}
```

With that present, the marked code would be replaced with this:
```go
for _, opt := range resolveRequestOptions(a.requestOptions, opts) {
	err = errors.Join(err, opt.ApplyToGet(&options))
}
```

File: pkg/api/types/options.go [hunk 8]
```diff
@@ -73,3 +107,26 @@ func (o *commonOptions) Set(key string, val interface{}, overwrite bool) error {
 	o.additional[key] = val
 	return nil
 }
+
+func (o commonOptions) GetEnvironment(key string) (string, error) {
```
Comment (LittleFox94) [inline pkg/api/types/options.go:111]: no doc comment..

File: pkg/api/errors.go [hunk 9]
```diff
@@ -22,6 +22,9 @@ var (
 
 	// ErrContextRequired is returned when a nil context was passed as argument.
 	ErrContextRequired = errors.New("no context given")
+
+	// ErrEnvironmentSegmentNotTypeString is returned when the environment segment is not of type string
+	ErrEnvironmentSegmentNotTypeString = errors.New("environment segment is not of type string")
```
Comment (LittleFox94) [inline pkg/api/errors.go:27]: not needed anymore, right?

**Review by 89Q12 — COMMENTED** [review]
LGTM, the failing integration tests well meh.
Otherwise doc strings have all been added.

**Comment by LittleFox94** [comment]
All relevant integration tests went through, I'm ok with merging without all of them being green
